infrastructure/deploy-sam-stack.sh

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
+set -e
+
+OPTION_REGEX="^--?.*"
+
+while [[ -n $1 ]]; do
+  case $1 in
+  -t | --template | --template-file)
+    shift
+    TEMPLATE=$1
+    ;;
+  -n | --stack-name)
+    shift
+    [[ -z ${1:-} || $1 =~ $OPTION_REGEX ]] && echo "Invalid stack name: '$1'" && exit 1
+    STACK_NAME=$1
+    ;;
+  -s3 | --prefix | --s3-prefix)
+    shift
+    PREFIX=$1
+    ;;
+  -g | --tags)
+    TAGS=()
+    while [[ $2 ]] && ! [[ $2 =~ $OPTION_REGEX ]]; do
+      shift
+      TAGS+=("$1")
+    done
+    ;;
+  -o | --params | --parameters | --parameter-overrides)
+    PARAMS=()
+    while [[ $2 ]] && ! [[ $2 =~ $OPTION_REGEX ]]; do
+      shift
+      PARAMS+=("$1")
+    done
+    ;;
+  -b | --build)
+    BUILD=true
+    ;;
+  -d | --no-build)
+    SKIP_BUILD=true
+    echo
+    ;;
+  -s | --base-dir)
+    shift
+    BASE_DIR=$1
+    ;;
+  -c | --cached)
+    CACHED=true
+    ;;
+  -p | --parallel)
+    PARALLEL=true
+    ;;
+  -y | --no-confirm)
+    NO_CONFIRM="--no-confirm-changeset"
+    ;;
+  *)
+    echo "Unknown option '$1'"
+    exit 1
+    ;;
+  esac
+  shift
+done
+
+[[ $TEMPLATE ]] && ! [[ -f $TEMPLATE ]] && echo "File '$TEMPLATE' does not exist" && exit 1
+[[ -z $STACK_NAME ]] && echo "Stack name not provided - not deploying (use the --stack-name|-n option)" && DEPLOY=false
+[[ $PREFIX ]] || PREFIX=$STACK_NAME
+${SKIP_BUILD:-false} && BUILD=false
+TAGS+=(DeploymentSource=Manual)
+
+: ${DEPLOY:=true}
+: ${BUILD:=false}
+
+if ! aws sts get-caller-identity > /dev/null 2>&1; then
+  echo "Valid AWS credentials not found in the environment. Authenticate to AWS before deploying a stack."
+  $DEPLOY && exit 1 || CREDS=false
+else
+  CREDS=true
+fi
+
+if $DEPLOY; then
+  echo "Deploying stack '$STACK_NAME'${TEMPLATE:+" from template '$TEMPLATE'"}"
+  echo "  with S3 prefix '$PREFIX'"
+  [[ ${#TAGS[@]} -gt 0 ]] && echo "  with tags '${TAGS[*]}'"
+  [[ ${#PARAMS[@]} -gt 0 ]] && echo "  with params '${PARAMS[*]}'"
+fi
+
+export AWS_DEFAULT_REGION=eu-west-2
+
+if [[ $TEMPLATE ]] && $BUILD && $CREDS; then
+  echo "Validating template..."
+  sam validate ${TEMPLATE:+--template "$TEMPLATE"}
+  sam validate ${TEMPLATE:+--template "$TEMPLATE"} --lint
```

Comment:
How does this exit if the template is invalid?

Reply:
How do you mean?
If the template file is specified but it doesn't exist, the script exits on line 63.
Actually it will always exist and be valid in this `if` block so the checks on lines 90-91 are not needed.

Reply:
`sam validate` is a good command to run, it's just that if the template is deemed to be invalid the script continues anyway.  I'd have thought a check on `$?` at the end of the if block would be useful so that it doesn't attempt to build and invalid template.

Reply:
I take it all back ... you've put set -e at the top ...